tests/gentropy/datasource/finngen/test_finngen_study_index.py

```diff
@@ -2,85 +2,343 @@
+        ("FinnGen r11", "Bleeding", "http://purl.obolibrary.org/obo/MP_0001914"),
+        # NOTE: Two EFO traits for one disease should be collected to array - hypothetical example:
+        # Glucose tolerance test & NMR Glucose
+        ("FinnGen r11", "Glucose", "http://www.ebi.ac.uk/efo/EFO_0002571"),
+        ("FinnGen r11", "Glucose", "http://www.ebi.ac.uk/efo/EFO_0004468"),
+        # NOTE: EFO that does not map, due to study not from Finngen - hypothetical example.
+        ("PheWAS 2024", "Glucose", "http://www.ebi.ac.uk/efo/EFO_0000001"),
     ]
-    curation_df = spark.createDataFrame(
-        data=curation_table_data,
-        schema=t.StructType(
-            [
-                t.StructField("STUDY", t.StringType(), nullable=False),
-                t.StructField("PROPERTY_VALUE", t.StringType(), nullable=False),
-                t.StructField("SEMANTIC_TAG", t.StringType(), nullable=False),
-            ]
-        ),
+    return data
+
+
+@pytest.fixture()
+def efo_mappings_df_mock(
+    spark: SparkSession, efo_mappings_mock: list[tuple[str, str, str]]
+) -> DataFrame:
+    """EFO mappings dataframe mock."""
+    schema = T.StructType(
+        [
+            T.StructField("STUDY", T.StringType(), nullable=False),
+            T.StructField("PROPERTY_VALUE", T.StringType(), nullable=False),
+            T.StructField("SEMANTIC_TAG", T.StringType(), nullable=False),
+        ]
     )
+    data = spark.createDataFrame(data=efo_mappings_mock, schema=schema)
+    return data
 
-    study_index = StudyIndex(_df=study_index_df, _schema=study_index_df.schema)
-    assert isinstance(
-        FinnGenStudyIndex.join_efo_mapping(
-            study_index,
-            finngen_release_prefix="FINNGEN_R11_",
-            efo_curation_mapping=curation_df,
-        ),
-        StudyIndex,
+
+@pytest.fixture()
+def urlopen_mock(
+    efo_mappings_mock: list[tuple[str, str, str, str]],
+    finngen_phenotype_table_mock: str,
+) -> Callable[[str], MagicMock]:
+    """Mock object for requesting urlopen objects with proper encoding.
+
+    This mock object allows to call `read` and `readlines` methods on two endpoints:
+    - https://finngen_phenotypes -> finngen_phenotype_table_mock
+    - https://efo_mappings -> efo_mappings_mock
+
+    The return values are mocks of the source data respectively.
+    """
+
+    def mock_response(url: str) -> MagicMock:
+        """Mock urllib.request.urlopen."""
+        match url:
+            case "https://finngen_phenotypes":
+                value = finngen_phenotype_table_mock
+            case "https://efo_mappings":
+                value = "\n".join(["\t".join(row) + "\n" for row in efo_mappings_mock])
+            case _:
+                value = ""
+        mock_open = MagicMock()
+        mock_open.read.return_value = value.encode()
+        mock_open.readlines.return_value = value.encode().splitlines(keepends=True)
+        return mock_open
+
+    return mock_response
+
+
+@pytest.mark.step_test
```

Comment:
I do not see the difference between them otherwise then the monkeypatch is more flexible and allows for mocking only part of the test. We can make it work with `patch` as well :)